### ISSUE TITLE
Work for API 0.7

### ISSUE THREAD

**Issue by divyanshumehta** [zerebubuth/openstreetmap-website#1]
Hey @zerebubuth I am also interested in working and getting the API 0.7 up for the openstreetmap-website. Can we discuss and make a plan. I have good experience with Rails APIs and I want to really do this task, together we can really get this up.

**Comment by zerebubuth** [comment]
Hi @divyanshumehta

Yeah, that sounds great!

Just a word of warning before we start: API 0.7 isn't something that's well-defined, and there are many different ideas about what it should contain. This is further complicated by the fact that we have two different API implementations (both the [Rails application](https://github.com/openstreetmap/openstreetmap-website) and a [CGI "accelerator"](https://github.com/zerebubuth/openstreetmap-cgimap)), which would either both need to implement API 0.7, or we'd need to retire one or the other. I had started work on [making the CGI functionality available to Ruby](https://github.com/zerebubuth/cgimap-ruby), but that has stalled and isn't anywhere close to being usable.

One thing that might be useful is to start separating out the parts of the Rails application which are "API" and which parts are "website", and make the website parts call the API. There are some things that the website does that the API currently isn't able to do (e.g: next/previous links) that we should extend the API to handle.

How does that sound?

**Comment by divyanshumehta** [comment]
Sound really cool, so basically we are trying to set up make the website get data by calling the API thereby separating the backend and the frontend of the website and then completely build on the new API. Also I think that a single implementation would be a better apporach for the new API.

Is the next and previous links , the browser buttons which appear on the website ?

Also we can have talks on IRC so as to how to get started, I am a bit of average, but I am really interested

**Comment by zerebubuth** [comment]
Yes, that's exactly what I mean. I think this is quite important to being able to develop API 0.7 independently of the Rails code (and vice-versa). I think @gravitystorm has a different opinion, though, so I'd like to hear from him.

Here's a conceptual diagram of how things are at the moment:

![image](https://user-images.githubusercontent.com/271360/36553258-c2e09cde-17f3-11e8-98aa-87eb1ac8dd40.png)

And, if we're able to make the Rails code call the API rather than go directly to the database, then:

![image](https://user-images.githubusercontent.com/271360/36553431-2abf0408-17f4-11e8-928e-452e2c70b4ff.png)

Which would allow us to swap out the implementation of the API, or move to different data stores, without being tied to how Rails represents data in a relational database:

![image](https://user-images.githubusercontent.com/271360/36553537-7ded64f8-17f4-11e8-91f7-dce5be083424.png)

The first step in that would be to make it possible (which it currently isn't) to run the whole website from the API. The next/previous links are one of the things that are currently missing. For example, the API doesn't return the information to put my next/previous changeset links at the bottom of [a changeset download page](https://www.openstreetmap.org/changeset/55416046#map=18/51.57672/-0.14177).

![image](https://user-images.githubusercontent.com/271360/36553783-2c4e6524-17f5-11e8-9338-865e8f6cef24.png)

Another example is the "also part of way..." data on shared nodes for the [way browse page](https://www.openstreetmap.org/way/56454731).

![image](https://user-images.githubusercontent.com/271360/36553935-9c3bcbec-17f5-11e8-96c3-e239a03a685d.png)

There are probably many other instances of this, but those are the two which came most directly to mind.

> Also we can have talks on IRC so as to how to get started

I hope you don't mind, but I'd prefer to keep technical discussion on Github. That means it's possible for other people to contribute to the discussion, and there's a place we can refer back to later on.

**Comment by divyanshumehta** [comment]
@zerebubuth was really helpful, excited for the work

**Comment by gravitystorm** [comment]
I think it's important to avoid mixing up two different concepts - api 0.7 is one thing, moving the website to use the API is another.

@zerebubuth is right, I have different opinions on moving the website to use the API (I don't support the idea) and I also have different opinions on moving all API functionality to cgimap (again, I don't support the idea). I would rather see cgimap embedded inside the rails code, to remove duplicated code but to ensure that we don't need to rewrite every single bit of the api in c++. For the website, I think simple html pages are best served by rails, rather than rails serving a javascript app which then makes api calls to just create html pages again. We could change the rails code to load these pages from a read-only connection, if there is actually a performance problem with the current setup.

**Comment by divyanshumehta** [comment]
@gravitystorm So it would be like the same controller returns html pages to web requests and JSON or XML to requests by the API. I guess this also a good option can be done using the 
```
respond_to do |format|
    format.html
    format.js
    format.xml
  end
````
How is this idea @zerebubuth @gravitystorm

**Comment by zerebubuth** [comment]
The existing Rails code has a [rather complex and non-standard way of writing output](https://github.com/openstreetmap/openstreetmap-website/blob/c32a63058c9caee42311a6ce5c0f88e5c31816b4/app/models/node.rb#L204-L218), which means it's sadly not as simple as `respond_to {|format| ... }`. This is for historical reasons, and dates back to a time when writing XML was hard because of the large size of the XML documents and the pressure this put on the Ruby garbage collector.

Something that I would find _very_ interesting would be to rewrite enough of the model output code to work with the more modern `builder` framework (or something else even more modern?), and checking to see what the performance difference is nowadays. It might be that we could make significant simplifications of the code for only modest performance losses.
